rllib/examples/metrics/custom_metrics_in_env_runners.py

```diff
@@ -0,0 +1,142 @@
+"""TODO (sven)
+"""
+import gymnasium as gym
+import numpy as np
+
+from ray.rllib.algorithms.callbacks import DefaultCallbacks
+from ray.rllib.env.wrappers.atari_wrappers import wrap_atari_for_new_api_stack
+from ray.rllib.utils.test_utils import (
+    add_rllib_example_script_args,
+    run_rllib_example_script_experiment,
+)
+from ray.tune.registry import get_trainable_cls, register_env
+
+
+class EnvRenderCallback(DefaultCallbacks):
+    """A custom callback to render the environment.
+
+    This can be used to create videos of the episodes for some or all EnvRunners
+    and some or all env indices (in a vectorized env). These videos can then
+    be sent to e.g. WandB as shown in this example script here.
+
+    We override the `on_episode_step` method to create a single ts render image
+    and temporarily store it in the Episode object.
+    """
+    def __init__(self):
+        super().__init__()
+        # Per sample round (on this EnvRunner), we want to only log the best- and
+        # worst performing episode's videos in the custom metrics. Otherwise, too much
+        # data would be sent to WandB.
+        self.best_episode_and_return = (None, float("-inf"))
+        self.worst_episode_and_return = (None, float("inf"))
+
+    def on_episode_step(
+        self,
+        *,
+        episode,
+        env_runner,
+        env,
+        env_index,
+        rl_module,
+        **kwargs,
+    ) -> None:
+        """Adds render image to episode."""
+        # If we have a vector env, only render the sub-env at index 0.
+        if isinstance(env.unwrapped, gym.vector.VectorEnv):
+            image = env.envs[0].render()
+        else:
+            image = env.render()
+        episode.add_temporary_timestep_data("render_images", image)
+
+    def on_episode_end(
+        self,
+        *,
+        episode,
+        env_runner,
+        env,
+        env_index,
+        rl_module,
+        **kwargs,
+    ) -> None:
+        # Get the episode's return.
+        episode_return = episode.get_return()
+
+        # Better than the best or worse than worst thus far?
+        if (
+            episode_return > self.best_episode_and_return[1]
+            or episode_return < self.worst_episode_and_return[1]
+        ):
+            # Get all images of the episode.
+            images = episode.get_temporary_timestep_data("render_images")
+            # Create a video from the images by simply stacking them.
+            video = np.stack(images, axis=0)
+
+            if episode_return > self.best_episode_and_return[1]:
+                self.best_episode_and_return = (video, episode_return)
+            else:
+                self.worst_episode_and_return = (video, episode_return)
+
+    def on_sample_end(
+        self,
+        *,
+        env_runner,
+        metrics_logger,
+        samples,
+        **kwargs,
+    ) -> None:
+        # Log the best and worst video to MetricsLogger.
+        env_runner.metrics.log_value(
```

Comment:
Dumb question: Where is this shown in TensorBoard/WandB? The structure of the metrics should be available somewhere. This becomes specifically important when running MARL (not necessarily for videos but for other agent-specific custom metrics).

Reply:
Good point. Maybe we should implement some nice print/visualization methods for the metrics class. Something that highlights the actual structure. Similar to `pprint`.